File: graphs/Clone Graph.py
from typing import Optional
from collections import deque

class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

def cloneGraph(node: Optional['Node']) -> Optional['Node']:
    if not node:
        return Node()

    old_to_new = {}
    queue = deque([node])
    visited = set()

    while queue:
        curr = queue.popleft()
        if curr in visited:
            continue
        visited.add(curr)

        if curr not in old_to_new:
            old_to_new[curr] = Node(curr.val)

        for n in curr.neighbors:
            if n not in old_to_new:
                old_to_new[n] = Node(n.val)
            old_to_new[curr].neighbors.append(old_to_new[n])
            queue.append(n)

    return old_to_new[node]

File: graphs/Clone Graph.cs
public class Node
{
    public int val;
    public IList<Node> neighbors;

    public Node()
    {
        val = 0;
        neighbors = new List<Node>();
    }

    public Node(int _val)
    {
        val = _val;
        neighbors = new List<Node>();
    }

    public Node(int _val, List<Node> _neighbors)
    {
        val = _val;
        neighbors = _neighbors;
    }
}


public class SolutionCloneGraph
{
    public Node CloneGraph(Node node)
    {
        if (node == null)
            return null;

        Dictionary<Node, Node> map = new Dictionary<Node, Node>();
        Queue<Node> queue = new Queue<Node>();
        queue.Enqueue(node);
        map[node] = new Node(node.val);

        while (queue.Count > 0)
        {
            Node curr = queue.Dequeue();

            foreach (Node n in curr.neighbors)
            {
                if (!map.ContainsKey(n))
                {
                    // create new child node if it does not exist and map it
                    map[n] = new Node(n.val);
                    queue.Enqueue(n);
                }

                // add neighbor to curr nodes neighbors list
                map[curr].neighbors.Add(map[n]);
            }
        }

        return map[node];
    }
}
